src/Modules/LearningTasks/Tutor.LearningTasks.Core/Domain/LearningTaskProgress/TaskProgress.cs

```diff
@@ -35,16 +38,80 @@ public void SubmitAnswer(int stepId, string answer)
     {
         var stepProgress = StepProgresses?.Find(s => s.StepId.Equals(stepId));
         stepProgress?.SubmitAnswer(answer);
+        Causes(new StepSubmitted(stepId));
 
         var allStepsAnswered = StepProgresses!.All(s => s.Status == StepStatus.Answered);
         if (allStepsAnswered)
+        {
             Status = TaskStatus.Completed;
+            Causes(new TaskCompleted());
+        }
     }
 
     public void ViewStep(int stepId)
     {
         var stepProgress = StepProgresses?.Find(s => s.StepId.Equals(stepId));
         stepProgress?.MarkAsViewed();
+
+        Causes(new StepOpened(stepId));
+        Causes(new SubmissionOpened(stepId));
+    }
+
+    public Result TaskOpened()
+    {
+        Causes(new TaskOpened());
+        return Result.Ok();
+    }
+
+    public Result OpenSubmission(int stepId)
+    {
+        Causes(new SubmissionOpened(stepId));
+        return Result.Ok();
+    }
+
+    public Result OpenGuidance(int stepId)
+    {
+        Causes(new GuidanceOpened(stepId));
+        return Result.Ok();
+    }
+
+    public Result OpenExample(int stepId)
+    {
+        Causes(new ExampleOpened(stepId));
+        return Result.Ok();
+    }
+
+    public Result PlayExampleVideo(int stepId)
+    {
+        Causes(new ExampleVideoPlayed(stepId));
+        return Result.Ok();
+    }
+
+    public Result PauseExampleVideo(int stepId)
+    {
+        Causes(new ExampleVideoPaused(stepId));
+        return Result.Ok();
+    }
+
+    public Result FinishExampleVideo(int stepId)
+    {
+        Causes(new ExampleVideoFinished(stepId));
+        return Result.Ok();
+    }
+
+    protected override void Apply(DomainEvent @event)
+    {
+        if (@event is not TaskProgresskEvent kcEvent) throw new EventSourcingException("Unexpected event type: " + @event.GetType());
+
+        kcEvent.LearningTaskId = LearningTaskId;
+        kcEvent.LearnerId = LearnerId;
+
+        When((dynamic)kcEvent);
+    }
+
+    private static void When(TaskProgresskEvent @event)
+    {
+        // No additional actions are required for all events.
```

Comment:
The `When` methods should change the state of the aggregate. At a minimum this means that `When(StepOpened)` and `When(StepSubmitted` occur we should invoke logic that is currently misplaced in the methods `SubmitAnswer` and `ViewStep`.

To summarize:
- In an event-source aggregate, methods called by services generate events,
- Event-handling methods change the aggregate's state.